### ISSUE TITLE
Crashing on Policy uploading

### ISSUE THREAD

**Issue by ChrOst** [sheagcraig/JSSImporter#36]
Hi,
as source for the crash see:
https://github.com/ChrOst/lrz-recipes/tree/master/addNetworkFavorite

What happens is the following:
```
Traceback (most recent call last):
  File "/usr/local/bin/autopkg", line 1469, in <module>
    sys.exit(main(sys.argv))
  File "/usr/local/bin/autopkg", line 1463, in main
    exit(subcommands[verb]['function'](argv))
  File "/usr/local/bin/autopkg", line 1287, in run_recipes
    autopackager.process(recipe)
  File "/Library/AutoPkg/autopkglib/__init__.py", line 466, in process
    self.env = processor.process()
  File "/Library/AutoPkg/autopkglib/__init__.py", line 295, in process
    self.main()
  File "/Library/AutoPkg/autopkglib/JSSImporter.py", line 729, in main
    self.policy = self.handle_policy()
  File "/Library/AutoPkg/autopkglib/JSSImporter.py", line 627, in handle_policy
    added_env="jss_policy_updated")
  File "/Library/AutoPkg/autopkglib/JSSImporter.py", line 525, in _update_or_create_new
    recipe_object.save()
  File "/Library/Python/2.7/site-packages/python_jss-0.5.8-py2.7.egg/jss/jss.py", line 1662, in save
    super(Policy, self).save()
  File "/Library/Python/2.7/site-packages/python_jss-0.5.8-py2.7.egg/jss/jss.py", line 631, in save
    "Object Conflict! If trying to post a new "
jss.exceptions.JSSPostError: Object Conflict! If trying to post a new object, look for name conflict and delete.

```
All in all I download a Release of a binary package, place it within a .pkg-root, and pkgize it.
Then it gets uploaded, but the policyupload crashes with an unknown error to me - maybe you can help.

Best regards,
Christoph

**Comment by ChrOst** [comment]
For reference - the errorcode from the JSS is this one:
409: There was a conflict when your request was processed. Normally this is due to your XML not including all of the required data, having invalid data or there is a conflict between what your resource and another one (e.g. some resources require a unique <name>). Check your XML and reattempt the request.

**Comment by sheagcraig** [comment]
Hi @ChrOst!
Based on the traceback, I agree-it's definitely failing at the policy creation level.

What I find when I see these 409/Object Conflict errors is usually some sneaky mismatch in object names.

The ```_update_or_create_new``` method referenced towards the bottom does just that-it looks to see whether an object (Policy in this case) already exists. If it does, it updates it, and if it doesn't, it makes a brand new one.

The conflict arises when it finds an existing one, but then tries to post an update that has some kind of conflict-like the ID in the URL doesn't match the ID in the XML being posted; I frequently have seen issues with groups where there's a difference between the input variable name for the group and what is specified in the template. It can also be a "conflict" in the sense that there's something incorrectly formed or specified in the XML. It could be a package name that doesn't actually exist, or differs from what the package is actually named, or a category, etc.

JAMF could do a lot better job reporting *what* the actual conflict is... It would make troubleshooting these things way easier. I may file a Feature Request to that effect today!

Regardless, nothing stands out as incorrect with your recipe and templates. It all looks great, actually. I've reviewed all of the files, and I even downloaded your recipe and ran it several times here to our testing server without any issues.

I can see that you've got the newest python-jss as well, which leads me to believe that you also have the newest JSSImporter. So I don't think updating either of those will really do anything.

Has this recipe successfully run at least once? Normally, this conflict occurs on the second and every subsequent run.

I guess the next thing to check would be to compare the XML of the existing policy with what JSSImporter is trying to update with and try to spot the conflicts.

I've looked at the one on my test JSS and not spotted anything, but maybe there's something difference from yours. (Which JSS version are you on?)

You could just grab the XML by using the policy GET on the api page, or if you're feeling up to just grabbing it with python-jss, that's an option as well.

**Comment by ChrOst** [comment]
Hi @sheagcraig 
Thanks for your quick message. :+1: 

Yes I got the updates you suggested last week - didn't update anything since that time.
The only difference may be that I have another Variable-set (like site_id/site_name) for Java installations, since java7 and java8 may not be installed at the same time. But that didnt have any effect on all other recipes so far. (I'll send you the diff)

There is no Policy which is named "addNetworkFavorite update" within my installation - even the Group gets created/updated properly. Since there is no policy i can't grab it :(

I'm on JSS 9.65.

It's not that urgent, since it would be an update Package only in the Future.
But the recipes/scripts pushed should work for other github release-binaries too, so that may be interesting :)

Christoph

**Comment by sheagcraig** [comment]
For the record, my test server is 9.65 as well, so I guess we can rule that out. I can't reproduce the issue here.

I guess additional steps would be to strip out parts of the recipe's JSSImporter arguments and elements of the policy template and see if it is possible to identify which element it is freaking out about.

Depending on how familiar you are with python-jss/python in general, it (may) be easier to use your policy template as a starting point to create a new Policy object in an interactive python interpreter to experiment with adding and removing elements, and see if you can recreate the issue there.

**Comment by ChrOst** [comment]
I'll do that later today. Will have a look into the database aswell, since there could be a Policy which isn't shown in the webui, which breaks this.

I'll post again if I have some news about it! Thanks so far

**Comment by ChrOst** [comment]
Hi @sheagcraig 
the crash reason here seems to be the Button Text, which was too long here (I'm not sure if thats JSS-Platform specific)

If you have filed the Feature Request, let me know, I'll vote it up! 

Thanks again :+1:

**Comment by sheagcraig** [comment]
Ha! I was thinking maybe the really-long name had something to do with it, but it didn't complain on our system. So strange.

I made a feature request, but then I took a closer look at things, and I think the errors *do* reference the problem (at least one problem, probably the first one encountered, at a time), so I feel like a doofus for putting the FR in. I'll add better logging/exception handling in the next version.
